### PR TITLE
enable nullable for commands

### PR DESCRIPTION
**Pull request by SimonCropp**

### CODE REVIEW

**Review by manfred-brands — CHANGES_REQUESTED** [review]
The original design of the protected field that needs to be set in derived classes is not the best and can be done better, either using a abstract method or a constructor parameter.

File: src/NUnitFramework/framework/Internal/Commands/BeforeAndAfterTestCommand.cs [hunk 1]
```diff
@@ -48,11 +48,11 @@ public override TestResult Execute(TestExecutionContext context)
...
         /// Perform the before test action
         /// </summary>
-        protected Action<TestExecutionContext> BeforeTest;
+        protected Action<TestExecutionContext> BeforeTest = null!;
 
         /// <summary>
         /// Perform the after test action
         /// </summary>
-        protected Action<TestExecutionContext> AfterTest;
+        protected Action<TestExecutionContext> AfterTest = null!;
```
[the remark this answers is not in the document]
Comment (SimonCropp) [reply src/NUnitFramework/framework/Internal/Commands/BeforeAndAfterTestCommand.cs:56]: wouldnt that be a breaking change?

[the remark this answers is not in the document]
Comment (manfred-brands) [reply src/NUnitFramework/framework/Internal/Commands/BeforeAndAfterTestCommand.cs:56]: Yes and even though the classes are in an internal namespace they are still public.

So let stick to exisiting design and simply declare the Action fields as nullable. They are already checked for null before use.

[the remark this answers is not in the document]
Comment (SimonCropp) [reply src/NUnitFramework/framework/Internal/Commands/BeforeAndAfterTestCommand.cs]: done

**Comment by SimonCropp** [comment]
this one should be good to go

**Comment by SimonCropp** [comment]
@manfred-brands 

> The original design of the protected field that needs to be set in derived classes is not the best and can be done better, either using a abstract method or a constructor parameter.

i would prefer to do that in another PR

**Comment by SimonCropp** [comment]
bump

File: src/NUnitFramework/framework/Internal/Commands/SetUpTearDownCommand.cs [hunk 2]
```diff
@@ -21,15 +22,9 @@ public SetUpTearDownCommand(TestCommand innerCommand, SetUpTearDownItem setUpTea
             Guard.OperationValid(Test.TypeInfo != null, "TestMethod must have a non-null TypeInfo");
             Guard.ArgumentNotNull(setUpTearDown, nameof(setUpTearDown));
 
-            BeforeTest = (context) =>
-            {
-                setUpTearDown.RunSetUp(context);
-            };
+            BeforeTest = setUpTearDown.RunSetUp;
```
Comment (stevenaw) [inline src/NUnitFramework/framework/Internal/Commands/SetUpTearDownCommand.cs:25]: 👍 for this simplification

File: src/NUnitFramework/framework/Internal/Commands/ConstructFixtureCommand.cs [hunk 3]
```diff
@@ -18,19 +20,14 @@ public ConstructFixtureCommand(TestCommand innerCommand)
...
+                if (typeInfo is {IsStaticClass: false})
                     // Use preconstructed fixture if available, otherwise construct it
-                    if (!typeInfo.IsStaticClass)
-                    {
-                        context.TestObject = Test.Fixture ?? typeInfo.Construct(((TestSuite)Test).Arguments);
-                        if (Test.Fixture == null)
-                            Test.Fixture = context.TestObject;
-                    }
+                {
+                    context.TestObject = Test.Fixture ??= typeInfo.Construct(((TestSuite)Test).Arguments);
```
Comment (stevenaw) [inline src/NUnitFramework/framework/Internal/Commands/ConstructFixtureCommand.cs:30]: It took me a bit to think through the possibilities here to ensure this was still the same logic but it seems sound

**Review by stevenaw — APPROVED** [review]
Thank you for this @SimonCropp !
LGTM but I'd appreciate a second pair of eyes for review before merging.

**Comment by SimonCropp** [comment]
@stevenaw who would be best to do that?

**Comment by stevenaw** [comment]
Oh, I see @manfred-brands already had a look.

**Pull request merged**